### ISSUE TITLE
Error installing from official repos onto Ubuntu

### ISSUE THREAD

**Issue by ganeshgunasekaran** [kubernetes/release#3219]
Hi,

I am tried installing the kubeadm in Ubuntu 22.4 LTS following the instructions given in the page https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/
I was able follow the steps and successfully installed Containerd .
The commands worked fine till Installing kubeadm -> Debian-based distributions -> Add the appropriate Kubernetes apt repository
The next step showed error while running "sudo apt-get update"

Please find the command outputs attached as screenshots Step 1-2 and step 3-4.

![step1-2](https://github.com/kubernetes/website/assets/87129782/9e322010-a6a9-4519-b9ea-8b158afd443e)
![step3-4](https://github.com/kubernetes/website/assets/87129782/a22c2d0b-2142-45cb-8a3e-4ce304635a6c)

This if my first issue. Please correct me if I have should have done something else.

**Comment by dominic-p** [comment]
For what it's worth, I'm also observing the "Access Denied" issue above. I'm not very familiar with how an APT repo is supposed to work, but most of the ones I have used allow browsing available packages (e.g. https://apt.kubernetes.io/).

**Comment by xmudrii** [comment]
@dominic-p What cloud provider or infrastructure are you using? For example, AWS tends to block access from some Hetzner IP addresses and we're aware of that, but there's nothing that we can do about that. Speaking of browsing the repo, this is not supported via `pkgs.k8s.io`, but you can do that via `download.opensuse.org`: https://download.opensuse.org/repositories/isv:/kubernetes:/

**Comment by dominic-p** [comment]
Thanks for the quick reply. That is very interesting as I am using Hetzner, and I've been struggling with network issues in a lot of areas. I assumed it was a misconfiguration on end, but maybe my IPs are blacklisted.

Just so I'm clear, it is expected behavior to get "Access Denied" when visiting https://pkgs.k8s.io/core:/stable:/v1.28/deb/ correct?

I originally came here because after switching from the Google repos to the community ones, my Hetzner load balancers no longer get any targets. The exact same config/version works when I install from the Google repos. I can open an issue on Hetzner's CCM, but I thought I would check here first. Are there any differences in the actual packages between the Google repos and the community ones?

**Comment by xmudrii** [comment]
> Just so I'm clear, it is expected behavior to get "Access Denied" when visiting https://pkgs.k8s.io/core:/stable:/v1.28/deb/ correct?

Yes, it's expected. We don't have a file browser at `pkgs.k8s.io`.

> Are there any differences in the actual packages between the Google repos and the community ones?

There should be no major differences. For both the Google repos and the community repos, we use the same binaries (e.g. `kubelet`, `kubeadm`...). Have you checked kubelet logs to make sure that kubelet is installed and running?

**Comment by dominic-p** [comment]
Thanks for the confirmation. Yes, the kubelet is running. I checked the logs, and I didn't see anything that looked out of the ordinary. It seems to have trouble connecting to the CRI-O unix socket for a bit at startup and then everything looks good. I'll take a look at the .deb files from each repo to see if I can see any differences.

**Comment by xmudrii** [comment]
@dominic-p Sounds good and please let us know if you find any difference. We tried our best to match these debs, and if there's any difference, it would be good to address it if it's possible.

**Comment by dominic-p** [comment]
Ok, I downloaded 1.28 kubelet debs from both the google repo and the community repo just now using a fresh Debian v12 container running on my cluster. I haven't looked deeply into the packages yet, but there are definitely some differences in the packages I downloaded.

```
$ tree
.
├── community
│   ├── etc
│   │   ├── kubernetes
│   │   │   └── manifests
│   │   └── sysconfig
│   │       └── kubelet
│   ├── lib
│   │   └── systemd
│   │       └── system
│   │           └── kubelet.service
│   ├── usr
│   │   ├── bin
│   │   │   └── kubelet
│   │   └── share
│   │       └── doc
│   │           └── kubelet
│   │               ├── LICENSE
│   │               └── README.md
│   └── var
│       └── lib
│           └── kubelet
└── google
    ├── lib
    │   └── systemd
    │       └── system
    │           └── kubelet.service
    └── usr
        └── bin
            └── kubelet
```

Here are the steps I did to get the debs:

### Community

1. Create new Debian 12 container
2. Run `apt update` followed by `apt install curl gpg`
3. Add community apt repo: 
```
curl -fsSL "https://pkgs.k8s.io/core:/stable:/v1.28/deb/Release.key" | gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg

echo "deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/v1.28/deb/ /" | tee /etc/apt/sources.list.d/kubernetes.list
```
4. Download kubelet: `apt download kubelet`: kubelet_1.28.1-1.1_amd64.deb

### Google

1. Create new Debian 12 container
2. Run `apt update` followed by `apt install curl gpg`
3. Add Google apt repo: 
```
curl -fsSL https://packages.cloud.google.com/apt/doc/apt-key.gpg | gpg --dearmor -o /etc/apt/keyrings/kubernetes-archive-keyring.gpg

echo "deb [signed-by=/etc/apt/keyrings/kubernetes-archive-keyring.gpg] https://apt.kubernetes.io/ kubernetes-xenial main" | tee /etc/apt/sources.list.d/kubernetes.list
```
4. Download kubelet: `apt download "kubelet=1.28.*"`: google-kubelet_1.28.1-00_amd64.deb

**Comment by xmudrii** [comment]
@dominic-p The new packages have additional files that we didn't have in packages published to the legacy Google-hosted repositories. It's important that we don't have missing files, i.e. that we installed some file with old packages, but that we don't do it with new packages.

**Comment by dominic-p** [comment]
Thanks for the explanation there. Ok, I think I was able to find the issue. The Hetzner CCM [currently requires](https://github.com/hetznercloud/hcloud-cloud-controller-manager/blob/v1.17.2/README.md#deployment) `--cloud-provider=external` to be set in `KUBELET_EXTRA_ARGS`. My configuration script sets the env variable in /etc/default/kubelet and the community package includes a new file /etc/sysconfig/kubelet with the contents `KUBELET_EXTRA_ARGS=`. That resets the env variable and breaks the Hetzner CCM.

I guess that makes this a bug with my particular configuration (I've already worked around it), but it is strange to me that the env variable is set in the new deb package when it wasn't before.

**Comment by xmudrii** [comment]
There's another report relevant to `/etc/default/kubelet`: https://github.com/kubernetes/release/issues/3276
We'll be taking a look into this as soon as possible.

**Comment by AnirudhPanchangam** [comment]
Hi Team,

I am experiencing a similar issue. 
Get:1 https://prod-cdn.packages.k8s.io/repositories/isv:/kubernetes:/core:/stable:/v1.28/deb  InRelease [1,186 B]
Err:1 https://prod-cdn.packages.k8s.io/repositories/isv:/kubernetes:/core:/stable:/v1.28/deb  InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 234654DA9A296436
Reading package lists... Done
W: GPG error: https://prod-cdn.packages.k8s.io/repositories/isv:/kubernetes:/core:/stable:/v1.28/deb  InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 234654DA9A296436
E: The repository 'https://pkgs.k8s.io/core:/stable:/v1.28/deb  InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.


I've run the same commands that OP has run. However, i get NO_PUBKEY issue. 
I tried running 
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 234654DA9A296436
Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).
Executing: /tmp/apt-key-gpghome.1PwUILV6CQ/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 234654DA9A296436
gpg: key 234654DA9A296436: public key "isv:kubernetes OBS Project <isv:kubernetes@build.opensuse.org>" imported
gpg: Total number processed: 1
gpg:               imported: 1

However, i still get the same error upon running `apt-get update`

Please let me know if there is anything else i can try.

Thanks,
Anirudh

**Comment by nethershaw** [comment]
I'm seeing this 403 response both from within AWS on EC2 instances and from my own system at home on a standard cable ISP. Same error from CloudFront.

I can see it's just CloudFront -> S3. Consider checking your bucket policy, and remember that S3 returns HTTP 403 not just for access denied, but also when it _would_ return HTTP 404. It does this on purpose so that unauthenticated users cannot use return codes to tell what files exist in the bucket. If you are looking for a permissions/policy problem, but it is actually a pathing problem, this behavior will conceal it.

**Comment by xmudrii** [comment]
@AnirudhPanchangam @nethershaw At this time, please use the official instructions for adding the repository (e.g. https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl).

**Comment by xmudrii** [comment]
Hey folks!

This issue now contains multiple distinct issues:

- 403 Forbidden trying to browse files: we don't have a file browser and the underlying S3 bucket is private, so accessing URLs of directories such as `https://pkgs.k8s.io/core:/stable:/v1.28/deb` is not going to work. #3317 has been created to track this
- 403 Forbidden trying to install packages: some cloud providers have IP addresses blocked by AWS. Hetzner is often affected by this. There is nothing that we can do about this, the same issue is affecting `registry.k8s.io`, and we can't make AWS to unblock the affected IP addresses. You can only try to get a new IP address that's not blocked or to mirror the repository somewhere else
- The issue with `/etc/default/kubelet` reported by @dominic-p is already fixed as part of #3279 
- For issues like `can't read signed-by key`, we strongly recommend setting up repositories as described in the official guidelines: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl
- About `Apt assumes the latter should exist when given https://pkgs.k8s.io/core:/stable:/v1.28/deb` reported by @nethershaw: we couldn't reproduce this issue, please try setting up the repository as described in the document that I linked previously. If the issue still appears, please create a new issue in this repository

Given that this issue contains multiple different reports and is hard to navigate, I'll go ahead and lock it. If you run into any issue and it's not already covered by this issue/comment, please create a new issue in this repository.
